core/src/nats.rs

```diff
@@ -12,19 +13,37 @@ use async_nats::{Client, Message, Subscriber};
 use bytes::Bytes;
 use dashmap::DashSet;
 use serde::{de::DeserializeOwned, Deserialize, Serialize};
-use std::error::Error;
 use std::fmt::Debug;
 use std::marker::PhantomData;
 use std::sync::Arc;
+use time::OffsetDateTime;
 use tokio_stream::StreamExt;
 
-use crate::logging::LogError;
-
 /// Unconstructable type, used as a [TypedMessage::Response] to indicate that
 /// no response is allowed.
 #[derive(Serialize, Deserialize)]
 pub enum NoReply {}
 
+pub struct MessageMeta {
+    pub timestamp: OffsetDateTime,
+    pub pending: u64,
+    pub sequence: u64,
+}
+
+impl TryFrom<jetstream::Message> for MessageMeta {
+    type Error = anyhow::Error;
+
+    fn try_from(value: jetstream::Message) -> Result<Self, Self::Error> {
+        let info = value.info().to_anyhow()?;
+
+        Ok(MessageMeta {
+            timestamp: info.published,
```

Comment:
whose time is this exactly? is it the time of the nats node at which the message is seen?

Reply:
what happens if the server republishes an old message?

Reply:
(Admittedly it's a little weird that this metadata comes in the form of the reply subject, but that's also where the API clients get it.)